c_src/erl_optics.c

```diff
@@ -46,14 +48,48 @@ static int load(ErlNifEnv *env, void **priv_data, ERL_NIF_TERM load_info) {
     atom_dist = enif_make_atom(env, "dist");
     atom_histo = enif_make_atom(env, "histo");
     atom_quantile = enif_make_atom(env, "quantile");
-    poller = optics_poller_alloc();
-    ctx = enif_alloc(sizeof(struct map_ctx));
-    if (poller) {
-      optics_poller_backend(poller, (void *) ctx, backend_eo, backend_free_eo);
-    }
     return 0;
 }
 
+static ERL_NIF_TERM eo_register_carbon_poller(
+    ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[])
+{
+    if (!poller){
+        poller = optics_poller_alloc();
+    }
+    if (!poller) return ERROR("poller_allocation_error");
+    if (carbon_poller) return ERROR("carbon_poller_already_registered");
+
+    enum {host_size = 255, port_size = 255};
+    char host[host_size] = {0};
+    char port[port_size] = {0};
+    if (enif_get_string(env, argv[1], host, host_size, ERL_NIF_LATIN1) < 1 ||
+        enif_get_string(env, argv[2], port, port_size, ERL_NIF_LATIN1) < 1) {
+        return ERROR("unable_to_decode(hostname|port)");
```

Comment:
I would recommend making two check so that your error message can easily tell which of the two parameters is wrong.